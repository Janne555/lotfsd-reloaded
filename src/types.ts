import { z } from "zod"

const AttributeSchema = z.object({
  value: z.number()
})

const SavingThrowSchema = z.object({
  value: z.number()
})

const AttackBonusSchema = z.object({
  value: z.number()
})

const HitPointsSchema = z.object({
  value: z.number()
})

const SurpriseChanceSchema = z.object({
  value: z.number().min(0).max(6),
})

const ArmorClassSchema = z.object({
  value: z.number()
})

const ActivitySchema = z.object({
  value: z.number().min(0).max(6),
})

const EncumbranceSchema = z.object({
  characterIsWearingChainMail: z.boolean(),
  characterIsWearingPlateMail: z.boolean()
})

const LanguageSchema = z.object({
  isKnown: z.boolean(),
  name: z.string(),
  id: z.string(),
})

const ItemSchema = z.object({
  name: z.string(),
  oversized: z.boolean(),
  id: z.string(),
})

const WeaponSchema = z.object({
  name: z.string(),
  attackBonus: z.number(),
  damage: z.string(),
  oversized: z.boolean(),
  id: z.string(),
  range: z
    .object({
      short: z.string().optional(),
      medium: z.string().optional(),
      long: z.string().optional(),
    })
    .optional(),
})

const SpellSchema = z.object({
  name: z.string(),
  description: z.string(),
  prepared: z.boolean(),
  level: z.number(),
  id: z.string(),
})

const CombatOptionSchema = z.object({
  name: z.string(),
})

const InfoSchema = z.object({
  name: z.string(),
  currentXp: z.number(),
  xpForNextLevel: z.number(),
  class: z.string(),
  race: z.string(),
  age: z.number(),
  gender: z.string(),
  alignment: z.string()
})

const AttributesSchema = z.object({
  charisma: AttributeSchema,
  constitution: AttributeSchema,
  dexterity: AttributeSchema,
  intelligence: AttributeSchema,
  strength: AttributeSchema,
  wisdom: AttributeSchema,
})

const SavingThrowsSchema = z.object({
  paralyze: SavingThrowSchema,
  poison: SavingThrowSchema,
  breathWeapon: SavingThrowSchema,
  magicalDevice: SavingThrowSchema,
  magic: SavingThrowSchema,
})

const CombatInfoSchema = z.object({
  baseAB: AttackBonusSchema,
  meleeAB: AttackBonusSchema,
  rangedAB: AttackBonusSchema,
  maxHP: HitPointsSchema,
  currentHP: HitPointsSchema,
  surpriseChance: SurpriseChanceSchema,
})

const ArmorClassesSchema = z.object({
  melee: ArmorClassSchema,
  ranged: ArmorClassSchema,
  withoutShield: ArmorClassSchema,
  surprised: ArmorClassSchema,
})

const ActivitiesSchema = z.record(ActivitySchema)

const EffectSchema = z.object({
  effect: z.enum(["add", "replace", "unknown"]),
  targetCategory: z.enum(["attributes", "attributeModifier", "savingThrows", "combatInfo", "armorClasses", "encumbrance"]),
  targetSubCategory: z.string(),
  valueNum: z.number().optional(),
  valueStr: z.string().optional(),
  id: z.string(),
  name: z.string(),
  active: z.boolean().optional(),
})

const SpellSlotSchema = z.object({
  level: z.number(),
  count: z.number(),
  usedCount: z.number()
})

const CharacterSheetSchema = z.object({
  id: z.string(),
  version: z.number(),
  info: InfoSchema,
  attributes: AttributesSchema,
  savingThrows: SavingThrowsSchema,
  combatInfo: CombatInfoSchema,
  armorClasses: ArmorClassesSchema,
  activities: ActivitiesSchema,
  encumbrance: EncumbranceSchema,
  languages: z.array(LanguageSchema),
  equipment: z.array(ItemSchema),
  nonEncumberingEquipment: z.array(ItemSchema),
  weapons: z.array(WeaponSchema),
  spells: z.array(SpellSchema),
  spellSlots: z.array(SpellSlotSchema),
  combatOptions: z.array(CombatOptionSchema),
  effects: z.array(EffectSchema)
})

export {
  AttributeSchema,
  SavingThrowSchema,
  AttackBonusSchema,
  HitPointsSchema,
  SurpriseChanceSchema,
  ArmorClassSchema,
  ActivitySchema,
  EncumbranceSchema,
  LanguageSchema,
  ItemSchema,
  WeaponSchema,
  SpellSchema,
  CombatOptionSchema,
  InfoSchema,
  AttributesSchema,
  SavingThrowsSchema,
  CombatInfoSchema,
  ArmorClassesSchema,
  ActivitiesSchema,
  CharacterSheetSchema,
  EffectSchema,
  SpellSlotSchema,
}

export type Attribute = z.infer<typeof AttributeSchema>
export type SavingThrow = z.infer<typeof SavingThrowSchema>
export type AttackBonus = z.infer<typeof AttackBonusSchema>
export type HitPoints = z.infer<typeof HitPointsSchema>
export type SurpriseChance = z.infer<typeof SurpriseChanceSchema>
export type ArmorClass = z.infer<typeof ArmorClassSchema>
export type Activity = z.infer<typeof ActivitySchema>
export type Encumbrance = z.infer<typeof EncumbranceSchema>
export type Language = z.infer<typeof LanguageSchema>
export type Item = z.infer<typeof ItemSchema>
export type Weapon = z.infer<typeof WeaponSchema>
export type Spell = z.infer<typeof SpellSchema>
export type CombatOption = z.infer<typeof CombatOptionSchema>
export type Info = z.infer<typeof InfoSchema>
export type Attributes = z.infer<typeof AttributesSchema>
export type SavingThrows = z.infer<typeof SavingThrowsSchema>
export type CombatInfo = z.infer<typeof CombatInfoSchema>
export type ArmorClasses = z.infer<typeof ArmorClassesSchema>
export type Activities = z.infer<typeof ActivitiesSchema>
export type CharacterSheet = z.infer<typeof CharacterSheetSchema>
export type Effect = z.infer<typeof EffectSchema>
export type SpellSlot = z.infer<typeof SpellSlotSchema>
