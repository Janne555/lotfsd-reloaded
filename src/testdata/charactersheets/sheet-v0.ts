export const CharacterSheetV0 = {
  id: 'v0-sheet',
  activities: [
    {
      id: '1',
      name: 'Climb',
      value: 3,
    },
    {
      id: '2',
      name: 'Stealth',
      value: 3,
    },
    {
      id: '3',
      name: 'Bushcraft',
      value: 3,
    },
    {
      id: '4',
      name: 'Languages',
      value: 3,
    },
    {
      id: '5',
      name: 'Open Doors',
      value: 3,
    },
    {
      id: '6',
      name: 'Search',
      value: 3,
    },
    {
      id: '7',
      name: 'Sneac Attack',
      value: 3,
    },
    {
      id: '8',
      name: 'Architecture',
      value: 3,
    }
  ],
  weapons: [
    {
      name: 'Sword',
      damage: '1d6',
      oversized: false,
      attackBonus: 1,
      id: '9',
    },
    {
      id: '10',
      name: 'Bow',
      damage: '1d6',
      range: {
        short: '1',
        medium: '2',
        long: '3',
      },
      oversized: false,
      attackBonus: 1,
    },
    {
      id: '11',
      name: 'Greatsword',
      damage: '1d10',
      oversized: true,
      attackBonus: 1,
    },
    {
      id: '12',
      name: 'Greatbow',
      damage: '1d10',
      range: {
        short: '1',
        medium: '2',
        long: '3',
      },
      oversized: true,
      attackBonus: 1,
    }
  ],
  equipment: [
    {
      id: '13',
      name: 'Rope',
      oversized: false,
    },
    {
      id: '14',
      name: 'Torch',
      oversized: false,
    },
    {
      id: '15',
      name: 'Lantern',
      oversized: false,
    },
    {
      id: '16',
      name: 'Backpack',
      oversized: false,
    },
    {
      id: '17',
      name: 'Candle',
      oversized: false,
    },
    {
      id: '18',
      name: 'Flint & Steel',
      oversized: false,
    },
    {
      id: '19',
      name: 'Waterskin',
      oversized: false,
    },
    {
      id: '20',
      name: 'Rations',
      oversized: false,
    },
    {
      id: '21',
      name: 'Bedroll',
      oversized: false,
    },
    {
      id: '22',
      name: 'Tent',
      oversized: false,
    },
    {
      id: '23',
      name: 'Crowbar',
      oversized: false,
    }
  ],
  info: {
    age: 20,
    alignment: 'Neutral',
    class: 'Fighter',
    currentXp: 0,
    xpForNextLevel: 1000,
    gender: 'Male',
    name: 'John Doe',
    race: 'Human'
  },
  attributes: {
    strength: {
      value: 10
    },
    dexterity: {
      value: 10
    },
    constitution: {
      value: 10
    },
    intelligence: {
      value: 10
    },
    wisdom: {
      value: 10
    },
    charisma: {
      value: 10
    },
  },
  effects: [
    {
      effect: 'add',
      targetCategory: 'attributes',
      targetSubCategory: 'strength',
      value: "1",
      id: '24',
      name: 'Strength +1',
    },
    {
      effect: 'add',
      targetCategory: 'attributeModifier',
      targetSubCategory: 'intelligence',
      value: "1",
      id: '25',
      name: 'Int +1',
    },
    {
      effect: 'add',
      targetCategory: 'savingThrows',
      targetSubCategory: 'paralyze',
      value: "1",
      id: '26',
      name: 'Paralyze +1',
    }
  ],
  savingThrows: {
    paralyze: {
      value: 10
    },
    poison: {
      value: 10
    },
    breathWeapon: {
      value: 10
    },
    magicalDevice: {
      value: 10
    },
    magic: {
      value: 10
    },
  },
  combatInfo: {
    baseAB: {
      value: 10
    },
    meleeAB: {
      value: 10
    },
    rangedAB: {
      value: 10
    },
    maxHP: {
      value: 10
    },
    currentHP: {
      value: 10
    },
    tempHP: {
      value: 10
    },
    surpriseChance: {
      value: 6
    },
  },
  armorClasses: {
    melee: {
      value: 12
    },
    ranged: {
      value: 12
    },
    withoutShield: {
      value: 12
    },
    surprised: {
      value: 12
    },
  },
  spells: [
    {
      name: 'Magic Missile',
      level: "1",
      description: 'A missile of magical energy darts forth from your fingertip and strikes its target, dealing 1d6+1 points of force damage.',
      id: 'spell-1'
    },
    {
      name: 'Fireball',
      level: "3",
      description: 'A fireball spell generates a searing explosion of flame that detonates with a low roar and deals 1d6 points of fire damage per caster level (maximum 10d6) to every creature within the area.',
      id: 'spell-2'
    },
    {
      name: 'Phantasamal Force',
      level: "2",
      description: 'You tap into the nightmares of a creature you can see within range and create an illusory manifestation of its deepest fears, visible only to that creature.',
      id: 'spell-3'
    }
  ],
  spellSlots: [
    {
      level: "1",
      preparedSpellId: 'spell-1',
      id: 'spell-slot-1',
      used: true
    },
    {
      level: "1",
      preparedSpellId: 'spell-1',
      id: 'spell-slot-2',
      used: false
    },
    {
      level: "1",
      id: 'spell-slot-3',
      used: false
    },
    {
      level: "2",
      preparedSpellId: 'spell-3',
      id: 'spell-slot-4',
      used: false
    },
    {
      level: "2",
      preparedSpellId: 'spell-3',
      id: 'spell-slot-5',
      used: false
    },
    {
      level: "3",
      preparedSpellId: 'spell-2',
      id: 'spell-slot-6',
      used: false
    },
    {
      level: "3",
      preparedSpellId: 'spell-2',
      id: 'spell-slot-7',
      used: false
    },
    {
      level: "3",
      id: 'spell-slot-8',
      used: false
    },
  ],
  languages: [
    {
      id: '28',
      name: 'Common',
      isKnown: true,
    },
    {
      id: '29',
      name: 'Kääpiötonttu',
      isKnown: false,
    },
    {
      id: '30',
      name: 'Käärmetonttu',
      isKnown: false,
    },
    {
      id: '31',
      name: 'Ancient common',
      isKnown: false,
    }
  ],
  nonEncumberingEquipment: [
    {
      id: '32',
      name: 'Locket',
      oversized: false
    }
  ],
  combatOptions: [
    {
      id: '33',
      name: 'Cleave',
    }
  ],
  encumbrance: {
    characterIsWearingChainMail: false,
    characterIsWearingPlateMail: false
  }
}