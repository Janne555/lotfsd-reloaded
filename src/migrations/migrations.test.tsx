import { describe, test, expect } from "vitest"
import { CharacterSheetV0 } from "../testdata/charactersheets/sheet-v0"
import { migrateDatabase } from "."
import { DatabaseSchema } from "../types"

describe("migrations", () => {
  test('should migrate from 0 to latest', async () => {
    const v0database = {
      version: 0,
      characterSheets: [
        CharacterSheetV0
      ]
    }

    const result = await migrateDatabase(v0database)

    expect(result.version).toBe(1)
    expect(() => DatabaseSchema.parse(result)).not.toThrow()
    expect(result).toMatchSnapshot()
  })
})
