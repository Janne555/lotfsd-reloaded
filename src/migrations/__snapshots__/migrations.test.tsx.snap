// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrations > should migrate from 0 to latest 1`] = `
{
  "characterSheets": [
    {
      "activities": [
        {
          "id": "1",
          "name": "Climb",
          "value": 3,
        },
        {
          "id": "2",
          "name": "Stealth",
          "value": 3,
        },
        {
          "id": "3",
          "name": "Bushcraft",
          "value": 3,
        },
        {
          "id": "4",
          "name": "Languages",
          "value": 3,
        },
        {
          "id": "5",
          "name": "Open Doors",
          "value": 3,
        },
        {
          "id": "6",
          "name": "Search",
          "value": 3,
        },
        {
          "id": "7",
          "name": "Sneac Attack",
          "value": 3,
        },
        {
          "id": "8",
          "name": "Architecture",
          "value": 3,
        },
      ],
      "armorClasses": {
        "melee": {
          "value": 12,
        },
        "ranged": {
          "value": 12,
        },
        "surprised": {
          "value": 12,
        },
        "withoutShield": {
          "value": 12,
        },
      },
      "attributes": {
        "charisma": {
          "value": 10,
        },
        "constitution": {
          "value": 10,
        },
        "dexterity": {
          "value": 10,
        },
        "intelligence": {
          "value": 10,
        },
        "strength": {
          "value": 10,
        },
        "wisdom": {
          "value": 10,
        },
      },
      "combatInfo": {
        "baseAB": {
          "value": 10,
        },
        "currentHP": {
          "value": 10,
        },
        "maxHP": {
          "value": 10,
        },
        "meleeAB": {
          "value": 10,
        },
        "rangedAB": {
          "value": 10,
        },
        "surpriseChance": {
          "value": 6,
        },
        "tempHP": {
          "value": 10,
        },
      },
      "combatOptions": [
        {
          "id": "33",
          "name": "Cleave",
        },
      ],
      "effects": [
        {
          "effect": "add",
          "id": "24",
          "name": "Strength +1",
          "targetCategory": "attributes",
          "targetSubCategory": "strength",
          "value": "1",
        },
        {
          "effect": "add",
          "id": "25",
          "name": "Int +1",
          "targetCategory": "attributeModifier",
          "targetSubCategory": "intelligence",
          "value": "1",
        },
        {
          "effect": "add",
          "id": "26",
          "name": "Paralyze +1",
          "targetCategory": "savingThrows",
          "targetSubCategory": "paralyze",
          "value": "1",
        },
      ],
      "encumbrance": {
        "characterIsWearingChainMail": false,
        "characterIsWearingPlateMail": false,
      },
      "equipment": [
        {
          "id": "13",
          "name": "Rope",
          "oversized": false,
        },
        {
          "id": "14",
          "name": "Torch",
          "oversized": false,
        },
        {
          "id": "15",
          "name": "Lantern",
          "oversized": false,
        },
        {
          "id": "16",
          "name": "Backpack",
          "oversized": false,
        },
        {
          "id": "17",
          "name": "Candle",
          "oversized": false,
        },
        {
          "id": "18",
          "name": "Flint & Steel",
          "oversized": false,
        },
        {
          "id": "19",
          "name": "Waterskin",
          "oversized": false,
        },
        {
          "id": "20",
          "name": "Rations",
          "oversized": false,
        },
        {
          "id": "21",
          "name": "Bedroll",
          "oversized": false,
        },
        {
          "id": "22",
          "name": "Tent",
          "oversized": false,
        },
        {
          "id": "23",
          "name": "Crowbar",
          "oversized": false,
        },
      ],
      "id": "v0-sheet",
      "info": {
        "age": 20,
        "alignment": "Neutral",
        "class": "Fighter",
        "currentXp": 0,
        "gender": "Male",
        "name": "John Doe",
        "race": "Human",
        "xpForNextLevel": 1000,
      },
      "languages": [
        {
          "id": "28",
          "isKnown": true,
          "name": "Common",
        },
        {
          "id": "29",
          "isKnown": false,
          "name": "Kääpiötonttu",
        },
        {
          "id": "30",
          "isKnown": false,
          "name": "Käärmetonttu",
        },
        {
          "id": "31",
          "isKnown": false,
          "name": "Ancient common",
        },
      ],
      "nonEncumberingEquipment": [
        {
          "id": "32",
          "name": "Locket",
          "oversized": false,
        },
      ],
      "savingThrows": {
        "breathWeapon": {
          "value": 10,
        },
        "magic": {
          "value": 10,
        },
        "magicalDevice": {
          "value": 10,
        },
        "paralyze": {
          "value": 10,
        },
        "poison": {
          "value": 10,
        },
      },
      "spellSlots": [
        {
          "id": "spell-slot-1",
          "level": "1",
          "preparedSpellId": "spell-1",
          "used": true,
        },
        {
          "id": "spell-slot-2",
          "level": "1",
          "preparedSpellId": "spell-1",
          "used": false,
        },
        {
          "id": "spell-slot-3",
          "level": "1",
          "used": false,
        },
        {
          "id": "spell-slot-4",
          "level": "2",
          "preparedSpellId": "spell-3",
          "used": false,
        },
        {
          "id": "spell-slot-5",
          "level": "2",
          "preparedSpellId": "spell-3",
          "used": false,
        },
        {
          "id": "spell-slot-6",
          "level": "3",
          "preparedSpellId": "spell-2",
          "used": false,
        },
        {
          "id": "spell-slot-7",
          "level": "3",
          "preparedSpellId": "spell-2",
          "used": false,
        },
        {
          "id": "spell-slot-8",
          "level": "3",
          "used": false,
        },
      ],
      "spells": [
        {
          "description": "A missile of magical energy darts forth from your fingertip and strikes its target, dealing 1d6+1 points of force damage.",
          "id": "spell-1",
          "level": "1",
          "name": "Magic Missile",
        },
        {
          "description": "A fireball spell generates a searing explosion of flame that detonates with a low roar and deals 1d6 points of fire damage per caster level (maximum 10d6) to every creature within the area.",
          "id": "spell-2",
          "level": "3",
          "name": "Fireball",
        },
        {
          "description": "You tap into the nightmares of a creature you can see within range and create an illusory manifestation of its deepest fears, visible only to that creature.",
          "id": "spell-3",
          "level": "2",
          "name": "Phantasamal Force",
        },
      ],
      "weapons": [
        {
          "attackBonus": 1,
          "damage": "1d6",
          "id": "9",
          "name": "Sword",
          "oversized": false,
        },
        {
          "attackBonus": 1,
          "damage": "1d6",
          "id": "10",
          "name": "Bow",
          "oversized": false,
          "range": {
            "long": "3",
            "medium": "2",
            "short": "1",
          },
        },
        {
          "attackBonus": 1,
          "damage": "1d10",
          "id": "11",
          "name": "Greatsword",
          "oversized": true,
        },
        {
          "attackBonus": 1,
          "damage": "1d10",
          "id": "12",
          "name": "Greatbow",
          "oversized": true,
          "range": {
            "long": "3",
            "medium": "2",
            "short": "1",
          },
        },
      ],
    },
  ],
  "version": 1,
}
`;
